### ISSUE TITLE
alingment between aimet-torch, aimet-onnx

### ISSUE THREAD

**Issue by taehokim3948** [quic/aimet#4088]
I found a significant mismatch between AIMET-Torch and AIMET-ONNX.

Specifically, I export an ONNX model and encodings from AIMET-Torch using sim.export(), then load them into AIMET-ONNX and run simulation. However, the outputs from AIMET-ONNX are very different from the original AIMET-Torch simulation results.

I expected the AIMET-ONNX simulation to be aligned with the AIMET-Torch simulation when using the exported ONNX model and encodings, but the difference is much larger than expected.

Could you help clarify whether this is expected behavior, or if there is something I may be missing in the export/load flow?

 AIMET-Torch and AIMET-ONNX: version 2.26.

SNR aimet-torch     vs   aimet-onnx  : 4.14 dB

export with aimet-torch 
`    ptq_sim = QuantizationSimModel(
        eng_model.cpu().eval(),
        dummy_input=to_cpu(tuple(decode_feed.values())),
        quant_scheme=QuantScheme.post_training_tf,
        default_param_bw=4,
        default_output_bw=16,
        in_place=True,
        config_file=str(args.aimet_config),
    )

    def forward_pass(model):
        with torch.no_grad():
            model(**decode_feed)

    ptq_sim.compute_encodings(forward_pass)

    cache.clear()
    decode_feed = get_decode_inputs(
        decode_input_ids, attention_mask,
        embed_layer, pos_gen, mask_gen, cache, device,
    )

    with torch.no_grad():
        qeng_out = ptq.sim.mode(**decode_feed)
    qeng_logits = np.array(qeng_out[0].cpu().detach())

    dummy_tuple = to_cpu(tuple(decode_feed.values()))
    ptq_sim.model = ptq_sim.model.cpu().eval()
    ptq_sim.export(
        dummy_input=dummy_tuple,
        path=str(decode_dir),
        filename_prefix="decode",
        onnx_export_args={
            "opset_version": 17,
            "input_names":  onnx_input_names(num_layers),
            "output_names": onnx_output_names(num_layers),
        },
    )
    print(f"[export] decode ONNX + encodings → {decode_dir}")`

load encodings and onnx with aimet-onnx 
`sim = QuantizationSimModel( model=model, dummy_input=create_dummy_input(model), quant_scheme=quant_scheme, param_type="int16", activation_type="int16", config_file=str(config_file), providers=["CUDAExecutionProvider", "CPUExecutionProvider"] if use_cuda else ["CPUExecutionProvider"], ) 

load_encodings_to_sim(sim, str(encodings_path), strict=False, allow_overwrite=False, disable_missing_quantizers=True)`

**Comment by quic-kyunggeu** [comment]
Hi @taehokim3948, thanks for the great question.

To get to the conclusion first, even without any bug, the **discrepancy between aimet-torch and aimet-onnx is normal**. The reason is fundamentally because aimet-onnx can quantize all activation (=non-static) tensors, but aimet-torch can't because of the dynamic nature of PyTorch. In practice, one of its most important implication is that aimet-torch can only quantize the input/output of nn.Modules (`nn.Softmax`) but not functional operators (`nn.functional.softmax`). Due to this limitation, you will often get a lot of missing activation encodings in the models exported from aimet-torch, and it's the main reason why we recommend aimet-onnx over aimet-torch by default to most users because with missing encodings, the model can frequently fall back to floating point kernels

Then why should one still use aimet-torch? The primary reason to use aimet-torch is quantization-aware training (QAT), which is impossible with ONNX models. In this scenario, you will
1) do QAT in aimet-torch (some of the activation quantizers will be missing, but let them be)
2) bring the model + **_partial_** encodings to aimet-onnx
3) let aimet-onnx quantsim fill out the missing activation quantizers, and
4) export the model + **_complete_** encodings for on-target deployment.

Having explained all that, I hope now I've also made it clear that in this scenario aimet-torch and aimet-onnx aren't necessarily meant to perfectly align with each other, but rather meant to be complimentary to each other. Aimet-torch specializes in QAT, and aimet-onnx fills out the missing encodings for optimal runtime performance.

**Comment by taehokim3948** [comment]
@quic-kyunggeu, Thank you again for the detailed explanation.

I am using AIMET for hardware that is not directly supported by AIMET. In my case, the ONNX model and partial encodings exported from AIMET-Torch are already sufficient for my hardware, without any missing parts on the deployment side.

What I need now is a way to validate that the exported ONNX model and encodings are correct before passing them to the compiler.

That is why I tried loading them with load_encodings_to_sim() in AIMET-ONNX. Since this function disables quantizers that are not present in the encodings, I expected the results to be reasonably close to the original AIMET-Torch simulation. However, in practice, the difference is still very large.

Could you help me understand why this gap is still so large in this case? Also, is there a recommended way to validate whether the exported ONNX model and encodings are correct before handing them off to the compiler?

**Comment by quic-kyunggeu** [comment]
I see. It's good that the encoding exported from aimet-torch is sufficient for your hardware. This will greatly simplify the your deployment process.

Even so, re-evaluating the aimet-torch's export artifacts (model + encoding) using aimet-onnx isn't the right way to validate the artifacts. As already explained, aimet-onnx almost always inserts more activation quantizer nodes than aimet-torch and produce different inference results. (Given that you said the encoding exported by aimet-torch was already sufficient in your case, aimet-onnx would insert more quantizers than you desire) The gap can be so large because some of the activations that weren't quantized in aimet-torch but were quantized in aimet-onnx might be very sensitive to quantization (extreme outliers, multimodal distribution, etc)

To validate aimet-torch's export artifacts, I recommend using our new export API [`aimet_torch.onnx.export`](https://quic.github.io/aimet-pages/releases/latest/apiref/torch/onnx.html#aimet-torch-onnx-export) instead. This takes the same list of input as torch.onnx.export and produce an "ONNX QDQ" model which you can run with onnxruntime.
<img width="912" height="289" alt="Image" src="https://github.com/user-attachments/assets/a8b49f4a-5c75-4f1f-ae6c-6cb38c0d83ed" />
With this, you can validate the soundness of the exported artifacts by comparing the output of onnxruntime and aimet-torch quantsim. If they don't match, please let us know. It's a P1 bug, and we will fix it asap.

Couple more tips about export
* [`aimet_torch.onnx.export` ](https://quic.github.io/aimet-pages/releases/latest/apiref/torch/onnx.html#apiref-torch-onnx) produces an onnx model in which represents the quantization encodings as Q/DQ nodes. If you want to store the quantization encodings into a separate JSON file, use [`sim.onnx.export`](https://quic.github.io/aimet-pages/releases/latest/apiref/torch/quantsim.html#aimet_torch.QuantizationSimModel.onnx) instead.
* `sim.export` in your example code is now deprecated and replaced by `aimet_torch.onnx.export` and `sim.onnx.export`. Going forward, we recommend using the new APIs. Although the old one may not cause any immediate problem in your case, it is known to crash or export incorrect encodings upon certain edge cases.

**Comment by taehokim3948** [comment]
@quic-kyunggeu, Thank you again for the detailed explanation and suggestion.

I tried `aimet_torch.onnx.export`, but I still observed that QDQ nodes are inserted automatically at additional locations, such as ouput of `Transpose`. From what I can see, this appears similar to the behavior of `sim.export`, in the sense that extra quantizers are still being added automatically in the exported ONNX graph. 

In my case, I would like to validate an ONNX graph that reflects the original AIMET-Torch quantization behavior as closely as possible, without introducing additional quantization points beyond what AIMET-Torch effectively used.

Is this automatic QDQ insertion expected behavior of `aimet_torch.onnx.export`? If so, is there any workaround or recommended way to prevent these extra QDQ nodes from being added around operators such as `Transpose`, so that the exported graph preserves only the original AIMET-Torch quantizers?
(only simulation with partial encodings without extra qdq node)

**Comment by taehokim3948** [comment]
@quic-kyunggeu, hi. I want to make sure whether my previous approach was incorrect.
I understand that load_encodings_to_sim() in AIMET-ONNX has a feature that disables quantizers that are not present in the encodings. Would it be incorrect to use this as a way to align AIMET-ONNX more closely with AIMET-Torch?

**Comment by Abdelmouhaimen** [comment]
> Hi [@taehokim3948](https://github.com/taehokim3948), thanks for the great question.
> 
> To get to the conclusion first, even without any bug, the **discrepancy between aimet-torch and aimet-onnx is normal**. The reason is fundamentally because aimet-onnx can quantize all activation (=non-static) tensors, but aimet-torch can't because of the dynamic nature of PyTorch. In practice, one of its most important implication is that aimet-torch can only quantize the input/output of nn.Modules (`nn.Softmax`) but not functional operators (`nn.functional.softmax`). Due to this limitation, you will often get a lot of missing activation encodings in the models exported from aimet-torch, and it's the main reason why we recommend aimet-onnx over aimet-torch by default to most users because with missing encodings, the model can frequently fall back to floating point kernels
> 
> Then why should one still use aimet-torch? The primary reason to use aimet-torch is quantization-aware training (QAT), which is impossible with ONNX models. In this scenario, you will
> 
> 1. do QAT in aimet-torch (some of the activation quantizers will be missing, but let them be)
> 2. bring the model + **_partial_** encodings to aimet-onnx
> 3. let aimet-onnx quantsim fill out the missing activation quantizers, and
> 4. export the model + **_complete_** encodings for on-target deployment.
> 
> Having explained all that, I hope now I've also made it clear that in this scenario aimet-torch and aimet-onnx aren't necessarily meant to perfectly align with each other, but rather meant to be complimentary to each other. Aimet-torch specializes in QAT, and aimet-onnx fills out the missing encodings for optimal runtime performance.

Hi, thanks for the detailed explanation.

I have a question regarding the workflow between aimet-torch and aimet-onnx.

I initially ran into issues with aimet-torch not generating encodings (or not quantizing) certain layers I needed. Because of that, I switched to aimet-onnx, where quantization works as expected and all layers get proper encodings.

However, I understand that QAT is only supported in aimet-torch, so I tried to go back to a Torch-based workflow. The problem I’m facing is:

The model calibrated in ONNX has different / more complete encodings (and sometimes different bitwidth assignments).
When I try to bring this model (or its encodings) back into aimet-torch for QAT, it doesn’t work directly because:
Some encodings are missing in Torch
Some quantizers differ (e.g., 8-bit vs 16-bit), causing incompatibilities
Torch cannot fully represent the same quantization structure as ONNX

So effectively, I’m stuck in this situation:

aimet-onnx → good coverage, but no QAT
aimet-torch → supports QAT, but incomplete quantization / encodings

**Comment by quic-kyunggeu** [comment]
> Is this automatic QDQ insertion expected behavior of aimet_torch.onnx.export? If so, is there any workaround or recommended way to prevent these extra QDQ nodes from being added around operators such as Transpose, so that the exported graph preserves only the original AIMET-Torch quantizers?
(only simulation with partial encodings without extra qdq node)

@taehokim3948 There is no way to suppress adding extra QDQ nodes in aimet_torch.onnx.export. The exported ONNX model with "extra" QDQ nodes is mathematically equivalent to the PyTorch quantsim, because the extra QDQ nodes are simply analytically derived from the adjacent encodings. (For example, `QDQ-Transpose` and `QDQ-Transpose-QDQ` produce same output, given the input and output QDQs share the same encoding)

> @quic-kyunggeu, hi. I want to make sure whether my previous approach was incorrect.
I understand that load_encodings_to_sim() in AIMET-ONNX has a feature that disables quantizers that are not present in the encodings. Would it be incorrect to use this as a way to align AIMET-ONNX more closely with AIMET-Torch?

That depends on your intention. If your goal was to make aimet-onnx **better** align with aimet-torch by inheriting the partial encodings, your approach is correct. However if your goal is making aimet-onnx **perfectly** aligned with aimet-torch, it is not always achievable, and there's no aimet api that guarantees that.


> I initially ran into issues with aimet-torch not generating encodings (or not quantizing) certain layers I needed. Because of that, I switched to aimet-onnx, where quantization works as expected and all layers get proper encodings.

@Abdelmouhaimen  Your observation is correct

> So effectively, I’m stuck in this situation:
>
> aimet-onnx → good coverage, but no QAT
aimet-torch → supports QAT, but incomplete quantization / encodings

Currently your order is the opposite. aimet-onnx -> aimet-torch is not a supported workflow.
Only aimet-torch -> aimet-onnx is supported.

Here's my recommendation step by step
1. Do QAT in aimet-torch, and export ONNX QDQ model with **partial/incomplete** encoding using [`aimet_torch.onnx.export`](https://quic.github.io/aimet-pages/releases/latest/apiref/torch/onnx.html#apiref-torch-onnx).
2. Take the exported model to aimet-onnx, using [`aimet_onnx.QuantizationSimModel.from_onnx_qdq`](https://quic.github.io/aimet-pages/releases/latest/apiref/onnx/quantsim.html#aimet_onnx.QuantizationSimModel.from_onnx_qdq).
(Recommend using aimet-onnx>=2.29)

**Comment by Abdelmouhaimen** [comment]
> Here's my recommendation step by step
> 
> 1. Do QAT in aimet-torch, and export ONNX QDQ model with **partial/incomplete** encoding using [`aimet_torch.onnx.export`](https://quic.github.io/aimet-pages/releases/latest/apiref/torch/onnx.html#apiref-torch-onnx).
> 2. Take the exported model to aimet-onnx, using [`aimet_onnx.QuantizationSimModel.from_onnx_qdq`](https://quic.github.io/aimet-pages/releases/2.28.0/apiref/onnx/quantsim.html#aimet_onnx.QuantizationSimModel.from_onnx_qdq).
>    (Recommend using aimet-onnx>=2.29)

Thanks, I managed to get past the previous issue, but I’m now hitting another problem when loading the exported encodings back into AIMET-ONNX.

The mismatch seems to be specifically related to mixed-precision bitwidths. Some layers that I configure as 16-bit in AIMET-ONNX are still exported from AIMET-Torch with 8-bit encodings. Then, when I call load_encodings_to_sim(...), I get a mismatch error because the quantizer in the ONNX sim is 16-bit while the loaded qdq is 8-bit.

So, if I understand correctly, compute_encodings(...) can fill in missing encodings, but it cannot resolve incompatible quantizer settings, such as a quantizer configured as 16-bit in the ONNX sim versus an 8-bit encoding coming from the QDQ ONNX exported after QAT.

Is that understanding correct?

**Comment by quic-kyunggeu** [comment]
> The mismatch seems to be specifically related to mixed-precision bitwidths. Some layers that I configure as 16-bit in AIMET-ONNX are still exported from AIMET-Torch with 8-bit encodings. Then, when I call load_encodings_to_sim(...), I get a mismatch error because the quantizer in the ONNX sim is 16-bit while the loaded qdq is 8-bit.

@Abdelmouhaimen Passing `strict=False` to `load_encodings_to_sim` should get you around the mismatch error. This will make your aimet-onnx sim's 16-bit quantizer be overwritten by the 8-bit encoding exported from aimet-torch. However, I would personally more recommend removing the 8-bit quantizer in aimet-torch in the beginning.

In your case, I think this should be closest to what you want:
1. Load partial encoding using `load_encodings_to_sim(..., strict=False, disable_missing_quantizers=False)`
2. Freeze all encodings that you **don't** want be overwritten, using `sim.qc_quantize_op_dict[<tensor_name>].freeze_encodings()`
3. Call `sim.compute_encodings`. This will only compute encodings of the quantizers that haven't been frozen in step 2.

For more information about the additional keyword argument, see our [API reference](https://quic.github.io/aimet-pages/releases/latest/techniques/ptq.html#api).

**Comment by bok3948** [comment]
@quic-kyunggeu Thank you for the details and your help. I will figure out how to solve my problem, but I really appreciate your help.
